lang/cs/Org.Apache.REEF.Client/Common/TempDirectory.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+// 
+//   http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.IO;
+
+namespace Org.Apache.REEF.Client.Common
+{
+    internal sealed class TempDirectory : IDisposable
+    {
+        private readonly string _directoryPath;
+        private readonly object _lock;
+        public TempDirectory(string directoryPath)
+        {
+            _directoryPath = directoryPath;
+            _lock = new object();
+        }
+
+        public void Dispose()
+        {
+            lock (_lock)
```

Comment:
Seems a bit strange for disposable to dispose of something it doesn't create. Could the object create the folder as well? i.e. modify `_driverFolderPreparationHelper.PrepareDriverFolder()` to return a new `TempDirectory` with the root directory created in the constructor.